anta/tools/utils.py

```diff
@@ -28,9 +29,72 @@ def get_failed_logs(expected_output: dict[Any, Any], actual_output: dict[Any, An
     for element, expected_data in expected_output.items():
         actual_data = actual_output.get(element)
 
+        if actual_data == expected_data:
+            continue
         if actual_data is None:
             failed_logs.append(f"\nExpected `{expected_data}` as the {element}, but it was not found in the actual output.")
         elif actual_data != expected_data:
             failed_logs.append(f"\nExpected `{expected_data}` as the {element}, but found `{actual_data}` instead.")
 
     return "".join(failed_logs)
+
+
+def custom_division(numerator: float, denominator: float) -> int | float:
+    """Get the custom division of numbers.
+
+    Custom division that returns an integer if the result is an integer, otherwise a float.
+
+    Args:
+    ----
+    numerator (float): The numerator.
+    denominator (float): The denominator.
+
+    Returns
+    -------
+    Union[int, float]: The result of the division.
+
+    """
+    result = numerator / denominator
+    return int(result) if result.is_integer() else result
+
+
+def extract_speed_and_lane(input_speed: str) -> tuple[str | None, int | None]:
+    """Get the speed and lane infromation.
+
+    Extracts the speed and lane information from the input string.
+
+    Args:
+    ----
+    input_speed (str): The input string which contains the speed and lane information.
+
+    Returns
+    -------
+    tuple[str|None, int|None]: The extracted speed from the input string, and the extracted lane from the input string.
+                                   If no lane information is found, it returns None.
+
+    Examples
+    --------
+        100g-8: (100, 8)
+        100g: (100, None)
+        auto: (None, None)
+        forced 100g: (100, None)
+        auto 100g: (100, None)
+        auto 100g-4: (100, 4)
+
+    """
+    # Regular expression pattern
+    # auto or forced: optional prefixes
+    # speed: a number (integer or float) followed by a mandatory 'g'
+    # lane: a number after a dash
+    pattern = r"(auto |forced )?(?P<speed>\d+(\.\d+)?)(g)(-(?P<lane>\d+))?"
```

Comment:
Why is this pattern different from InterfaceSpeedPAttern

```
r"^((auto)?\s?(\d{1,4}(\.\d{1})?(g)?(-\d{1,2})?)?|forced\s\d{1,4}(g)?)$"
```

?

Reply:
InterfaceSpeedPAttern is used to restrict the speed as auto and forced have some difference in type of speeds.
Example:
Forced not support lane and 2.5g speed but auto support the same so made the pattern accordingly.

About the pattern to collect the speed and lane we wont need to define same as InterfaceSpeedPAttern because here we are just collecting the speed and lane from valid input. 
If we are going to define this pattern same as InterfaceSpeedPAttern then we need to use one more key as `?P<forced_speed>`.

Reply:
1/ good for making the diff in input

I an wondering if our problem is not simpky that we are not modelling well enough all of this


2/